morpheus/_lib/src/stages/filter_detection.cpp

```diff
@@ -33,126 +36,193 @@
 #include <cstdint>  // for uint8_t
 #include <exception>
 #include <memory>
-#include <ostream>      // needed for glog
+#include <ostream>  // needed for glog
+#include <string>
 #include <type_traits>  // for declval (indirectly via templates)
 #include <utility>      // for pair
 // IWYU thinks we need ext/new_allocator.h for size_t for some reason
 // IWYU pragma: no_include <ext/new_allocator.h>
 
+namespace {
+
+/**
+ * @brief Simple container describing a 2d buffer
+ *
+ * TODO: Merge with DevMemInfo?
+ */
+struct BufferInfo
+{
+    std::vector<std::size_t> shape;
+    std::vector<std::size_t> stride;
+    std::size_t count;
+    std::size_t bytes;
+    morpheus::DType type;
+    const uint8_t* head;
+};
```

Comment:
@mdemoret-nv I wasn't sure if I should just add shape, stride & bytes to the `DevMemInfo` struct or not. 

On one hand I didn't want to make this PR any larger than it already is. However on the other hand, without these fields `DevMemInfo` many of matx_util methods which also require passing in the shape and stride.

Reply:
I think this is a good time to merge them. Keeping it separate will only lead to more work down the line. If we go down this route, we probably want to make a few changes:

1. Unify the `data` property. One uses a pointer and the other uses a `rmm::device_buffer`
2. We probably want to hold a reference to the `rmm::device_buffer` but allow a `data()` method
3. Most of these fields should be const so they dont get modified after being set. `DevMemInfo` should be immutable
4. We will probably want some utility methods to simplify getting the bytes per element and things like that.